From c497ad172b6c0db51dad79f71e55b99010164557 Mon Sep 17 00:00:00 2001
From: Antonio Rojas <arojas@archlinux.org>
Date: Sat, 24 May 2025 15:19:29 +0200
Subject: [PATCH] Drop wrapperdescr_fastcall usage for Cython 3.1 compatibility

In Cython 3.1 binary operations are no longer wrappers, but methods. This breaks wrapperdescr_fastcall.

Since this was rather hackish anyway, we replace its few uses by explicit coercions.
---
 .../known-test-failures.json                  |   4 -
 src/doc/en/reference/cpython/index.rst        |   1 -
 .../lie_algebras/lie_algebra_element.pyx      |  45 +++++++-
 src/sage/cpython/meson.build                  |   1 -
 src/sage/cpython/wrapperdescr.pyx             | 104 ------------------
 .../rings/polynomial/polynomial_element.pyx   |  15 ++-
 6 files changed, 48 insertions(+), 122 deletions(-)
 delete mode 100644 src/sage/cpython/wrapperdescr.pyx

Index: sage-test/sage/src/doc/en/reference/cpython/index.rst
===================================================================
--- sage-test.orig/sage/src/doc/en/reference/cpython/index.rst
+++ sage-test/sage/src/doc/en/reference/cpython/index.rst
@@ -12,7 +12,6 @@ internals.
    sage/cpython/debug
    sage/cpython/getattr
    sage/cpython/cython_metaclass
-   sage/cpython/wrapperdescr
    sage/cpython/dict_del_by_value
 
 .. include:: ../footer.txt
Index: sage-test/sage/src/sage/algebras/lie_algebras/lie_algebra_element.pyx
===================================================================
--- sage-test.orig/sage/src/sage/algebras/lie_algebras/lie_algebra_element.pyx
+++ sage-test/sage/src/sage/algebras/lie_algebras/lie_algebra_element.pyx
@@ -24,7 +24,6 @@ from sage.misc.repr import repr_lincomb
 from sage.combinat.free_module import CombinatorialFreeModule
 from sage.structure.element cimport have_same_parent, parent
 from sage.structure.coerce cimport coercion_model
-from sage.cpython.wrapperdescr cimport wrapperdescr_fastcall
 from sage.structure.element_wrapper cimport ElementWrapper
 from sage.structure.richcmp cimport richcmp, richcmp_not_equal
 from sage.data_structures.blas_dict cimport axpy, add, negate, scal
@@ -68,11 +67,28 @@ cdef class LieAlgebraElement(IndexedFree
         """
         try:
             # Try the normal coercion first
-            return wrapperdescr_fastcall(IndexedFreeModuleElement.__mul__,
-                                         left, (right,), <object>NULL)
+            return IndexedFreeModuleElement.__mul__(left, right)
         except TypeError:
             pass
 
+        try:
+            # Handle the case of right multiplication by scalar
+            if isinstance(left, IndexedFreeModuleElement):
+                R = (<IndexedFreeModuleElement>left)._parent._base
+                x = R.coerce(right)
+                return IndexedFreeModuleElement.__mul__(left, x)
+        except (TypeError, KeyError):
+            pass
+
+        try:
+            # Handle the case of left multiplication by scalar
+            if isinstance(right, IndexedFreeModuleElement):
+                R = (<IndexedFreeModuleElement>right)._parent._base
+                x = R.coerce(left)
+                return IndexedFreeModuleElement.__mul__(x, right)
+        except (TypeError, KeyError):
+            pass
+
         # Lift up to the UEA and try multiplication there
         # We will eventually want to lift stuff up anyways,
         #   so just do it here.
@@ -348,9 +364,26 @@ cdef class LieAlgebraElementWrapper(Elem
         """
         try:
             # Try the normal coercion first
-            return wrapperdescr_fastcall(ElementWrapper.__mul__,
-                                         left, (right,), <object>NULL)
-        except TypeError:
+            return ElementWrapper.__mul__(left, right)
+        except TypeError as e:
+            pass
+
+        try:
+            # Handle the case of right multiplication by scalar
+            if isinstance(left, LieAlgebraElementWrapper):
+                R = (<LieAlgebraElementWrapper>left)._parent._base
+                x = R.coerce(right)
+                return ElementWrapper.__mul__(left, x)
+        except (TypeError, KeyError):
+            pass
+
+        try:
+            # Handle the case of left multiplication by scalar
+            if isinstance(right, LieAlgebraElementWrapper):
+                R = (<LieAlgebraElementWrapper>right)._parent._base
+                x = R.coerce(left)
+                return ElementWrapper.__mul__(x, right)
+        except (TypeError, KeyError):
             pass
 
         # Lift up to the UEA and try multiplication there
Index: sage-test/sage/src/sage/rings/polynomial/polynomial_element.pyx
===================================================================
--- sage-test.orig/sage/src/sage/rings/polynomial/polynomial_element.pyx
+++ sage-test/sage/src/sage/rings/polynomial/polynomial_element.pyx
@@ -61,7 +61,6 @@ import operator
 import copy
 import re
 
-from sage.cpython.wrapperdescr cimport wrapperdescr_fastcall
 import sage.rings.rational
 import sage.rings.integer
 from . import polynomial_ring
@@ -2217,11 +2216,15 @@ cdef class Polynomial(CommutativeAlgebra
             except TypeError:
                 pass
 
-        # Delegate to coercion model. The line below is basically
-        # RingElement.__truediv__(left, right), except that it also
-        # works if left is not of type RingElement.
-        return wrapperdescr_fastcall(RingElement.__truediv__,
-                left, (right,), <object>NULL)
+        if isinstance(right, Polynomial):
+            R = (<Polynomial>right)._parent
+            try:
+                x = R.coerce(left)
+                return x.__truediv__(right)
+            except TypeError:
+                pass
+
+        return NotImplemented
 
     def __pow__(left, right, modulus):
         """
